src/Analyzer/Passes/LogicalExpressionOptimizerPass.cpp

```diff
@@ -1130,19 +1130,25 @@ class LogicalExpressionOptimizerVisitor : public InDepthQueryTreeVisitorWithCont
 
         /// To avoid duplicate when traversing in equal condition
         std::unordered_set<QueryTreeNodePtr> equal_set;
+        /// To avoid endless loop, for example, a>b AND b>a AND a<5
+        std::unordered_set<QueryTreeNodePtr> check_path;
```

Comment:
looks like `equal_set` and `check_path` can be merged together

Reply:
They are serving for different purposes. For functionality, looks like shouldn't be merged; For clarity, I would also differentiate them.

Reply:
I cannot agree they are different. All the conditions, no matter `> < =`, all lay in `greater pairs` or/and `less pairs`. We can use a single set called `visited_node` to avoid cyclic walk in DFS. It's just a graph, and there is nothing special for the edges.

Reply:
They mainly differ in this [line of code](https://github.com/ClickHouse/ClickHouse/blob/94607b9714921c092187d727d2e65f06c70289c5/src/Analyzer/Passes/LogicalExpressionOptimizerPass.cpp#L1209). I added a [testcase](https://github.com/ClickHouse/ClickHouse/blob/94607b9714921c092187d727d2e65f06c70289c5/tests/queries/0_stateless/03285_constant_compare_in_and_chain.sql#L53) to show the difference. If we merge them together, we would miss the case of `c<5`.

Reply:
Hi, I think it's same for equals, I tried
```
a = b AND a > 3 AND b > 0
```
and got
```
(a = b) AND (a > 3) AND (b > 0) AND (b > 3)
```
We don't have `a > 0` here :)

Reply:
Here `check_path.clear(); equal_set.clear();` should be in the loop.